staging/src/k8s.io/apiserver/pkg/admission/plugin/cel/evaluator.go

```diff
@@ -0,0 +1,237 @@
+	}
+	return NewEvaluator(compilationResults, validations)
+}
+
+// evaluator implements the Evaluator interface
+type evaluator struct {
+	compilationResults []CompilationResult
+	validations        []Validation
+}
+
+func NewEvaluator(compilationResults []CompilationResult, validations []Validation) Evaluator {
+	return &evaluator{
+		compilationResults,
+		validations,
+	}
+}
+
+func convertObjectToUnstructured(obj interface{}) (*unstructured.Unstructured, error) {
+	if obj == nil || reflect.ValueOf(obj).IsNil() {
+		return &unstructured.Unstructured{Object: nil}, nil
+	}
+	ret, err := runtime.DefaultUnstructuredConverter.ToUnstructured(obj)
+	if err != nil {
+		return nil, err
+	}
+	return &unstructured.Unstructured{Object: ret}, nil
+}
+
+func objectToResolveVal(r runtime.Object) (interface{}, error) {
+	if r == nil || reflect.ValueOf(r).IsNil() {
+		return nil, nil
+	}
+	v, err := convertObjectToUnstructured(r)
+	if err != nil {
+		return nil, err
+	}
+	return v.Object, nil
+}
+
+// Evaluate evaluates the compiled CEL expressions converting them into CELEvaluations
+// errors per evaluation are returned on the Evaluation object
+func (v *evaluator) Evaluate(versionedAttr *generic.VersionedAttributes, versionedParams runtime.Object) ([]Evaluation, error) {
+	// TODO: replace unstructured with ref.Val for CEL variables when native type support is available
+	evaluations := make([]Evaluation, len(v.compilationResults))
+	var err error
+
+	oldObjectVal, err := objectToResolveVal(versionedAttr.VersionedOldObject)
+	if err != nil {
+		return nil, err
+	}
+	objectVal, err := objectToResolveVal(versionedAttr.VersionedObject)
+	if err != nil {
+		return nil, err
+	}
+	paramsVal, err := objectToResolveVal(versionedParams)
+	if err != nil {
+		return nil, err
+	}
+
+	request := createAdmissionRequest(versionedAttr.Attributes)
+	requestVal, err := convertObjectToUnstructured(request)
+	if err != nil {
+		return nil, err
+	}
+	va := &validationActivation{
+		object:    objectVal,
+		oldObject: oldObjectVal,
+		params:    paramsVal,
+		request:   requestVal.Object,
+	}
+
+	for i, compilationResult := range v.compilationResults {
+		var evaluation = &evaluations[i]
+		validation := v.validations[i]
+
+		evaluation.Validation = validation
+		if compilationResult.Error != nil {
+			evaluation.Error = errors.New(fmt.Sprintf("compilation error: %v", compilationResult.Error))
+			continue
+		}
+		if compilationResult.Program == nil {
+			evaluation.Error = errors.New("unexpected internal error compiling expression")
+			continue
+		}
+		t1 := time.Now()
+		evalResult, _, err := compilationResult.Program.Eval(va)
+		elapsed := time.Since(t1)
+		evaluation.Elapsed = elapsed
+		if err != nil {
+			evaluation.Error = errors.New(fmt.Sprintf("expression '%v' resulted in error: %v", validation.Expression, err))
+		} else {
+			evaluation.EvalResult = evalResult
+		}
+	}
+
+	return evaluations, nil
```

Comment:
`Evaluations` is a list of raw evaluation results. For both evaluation and matching, we know that evaluations should be interpreted as the AND-ing of matching.

evaluation is also interested in building a response using Message strings.

It seems to me a key part of the interface is figuring out how we expect the caller to interpret the results. I have some thoughts on that in my comments on the `interface.go` file for this package.

Reply:
evaluation is now only related to cel 

```
// EvaluationResult contains the minimal required fields and metadata of a cel evaluation
type EvaluationResult struct {
	EvalResult ref.Val
	Expression string
	Elapsed    time.Duration
	Error      error
}
```